s02/2.md

```diff
@@ -10,14 +10,556 @@ Tartalom
+        ~ Array ();
+    private:
+        double *val;
+};
+Array operator+(const Array& left,const Array& right){
+    Array res = left;
+    res += right;
+    return res;
+}
+```
+
+Az `Array` egy osztály, melynek + operátora összekonkatenálja a két paramétert és visszaad egy új listát.
+
+A következő függvény meghívásánál azonban több köztes `Array` példány keletkezik és szűnik meg:
+
+```C++
+void f()
+{
+ Array b, c, d;
+ …
+ Array a = b + c + d;
+}
+```
+
+A move-szemantika az ehhez hasonló problémákra ad megoldást.
+
+* másolás helyett "ellopja" az erőforrást
+* törölhető állapotban hagyja a másik objektumot
+
+Ehhez kell:
+
+* overloadolni lehessen a _Copy-konstruktort_ és az _értékadó operátort_, illetve egyéb függvényeket
+    + meg kell tartani a backward compatibility-t
+    + meg kell különböztetni a bal- és jobbértékeket
+
+#### RValue, LValue
+
+Korábbi nyelvekben értékadás: `<variable> = <expression>` (pl.: `x = 5`)
+
+C/C++-ban értékadás: `<expression> = <expression>` (pl.: `*++ptr = *++qtr`)
+
+* de nem minden esetben működik, pl.: `a+5 = x` helytelen.
+
+**LValue**: kifejezés, mely egy memóriaterületre hivatkozik és lehetővé teszi, hogy a `&` operátorral
+megszerezzük annak memóriacímét
+
+**RValue**: olyan kifejezés, mely nem LValue.
+
+#### Move-operátor
+
+A move-operátorral castolni lehet LValue-t RValue-vá. (balértékből jobbértéket)
+
+* kikényszeríthethük a move-szemantika használatát
+
+Példa: 
+```C++
+struct S
+{
+ S() { a = ++cnt; std::cout << “S() ”; }
+ S(const S& rhs) { a = rhs.a; std::cout << “copyCtr ”; }
+ S(S&& rhs) { a = rhs.a; std::cout << “moveCtr ”; }
+ S& operator=(const S& rhs) { a = rhs.a; std::cout << “copy= ”; return *this; }
+ S& operator=(S&& rhs) { a = rhs.a; std::cout << “move= ”; return *this; }
+ int a ;
+ static int cnt;
+};
+int S::cnt = 0;
+
+int main()
+{
+ S a, b;
+ swap( a, b);
+}
+```
+
+Move-operátor használata nélkül:
+
+```C++
+template<class T>
+void swap(T& a, T& b)
+{
+    T tmp(a);
+    a = b;
+    b = tmp;
+}
+
+```
+
+Eredmény: `S() S() copyCtr copy= copy=`
+
+Move-operátor használatával:
+
+```C++
+template<class T>
+void swap(T& a, T& b)
+{
+    T tmp(std::move(a));
+    a = std::move(b);
+    b = std::move(tmp);
+}
+```
+
+Eredmény: `S() S() moveCtr move= move=`
+
 ## 3. Erőforrásbiztos programozás, RAII, destruktor és szemétgyűjtés <a name="chapter03"></a>
 
 ## 4. Kivételkezelés, kivételbiztos programozás <a name="chapter04"></a>
 
+### Hibakezelés
+
+Hiba: Eljárás előfeltételének megsértése miatt, vagy bármely más nem kívánt működésből adódó állapot. 
+
+#### Hibakezelés klasszikus C-ben
+
+* függvény visszatérési értékei hibakódok / NULL referencia
+* externális `errno` kód
+* visszatérési érték vizsgálat, pl.: `fseek`, `int`-et ad, `bool`-ként kezeljük
+* hiba flag-ek, pl.: hibaflag beállítódik $\Rightarrow$ többi IO művelet nem csinál semmit
+* `assert()`: nem teljesül a feltétel $\Rightarrow$ hibaüzenet, mely a felhasználó számára nemigazán érthető
+ 
```

Comment:
\+

* hibakód változójának paraméterként való átadása referenciaként (klasszikus C-ben pointer)
* struct-al való visszatérés, mely az eredmény mellett tartalmazza a hibakódot is (mintha egy pair-ünk lenne)